enhancements/microshift/multus-cni-for-microshift.md

```diff
@@ -0,0 +1,401 @@
+Multus CNI binary (which is copied to host's `/opt/cni/bin`).
+
+To supply network plugins (CNIs) such as bridge, vlan, macvlan, and so on a new image will be prepared
+so MicroShift uses the same binaries as OpenShift (alternative is using RHEL's networkplugins RPM - see alternatives).
+This image will also contain IPAM binaries such as `static`, `dynamic (DHCP)`, and `host-local`.
+
+#### Hypershift [optional]
+
+No, enhancement is MicroShift specific.
+
+### Risks and Mitigations
+
+Taking `bridge` CNI as an example: when the bridge interface does not exist, it will be created
+when a Pod requiring that interface is created.
+If user expects something else will create the interface, they will need to configure system to start
+MicroShift after other services.
+
+Multus using thin plugin architecture creates a kubeconfig when the DaemonSet starts and copies it
+out of Pod to location on disk so the Multus CNI binary can use it to get NetworkAttachmentDefinitions
+and Pod annotations. The file is owned and readable only by root, so the risk of someone gaining
+access to the cluster after logging into the host can be compared to getting access to
+`/var/lib/microshift/resources/kubeadmin/kubeconfig`.
+
+Currently MicroShift runs as a single-node cluster but in future releases there might be an effort
+to allow for multi-node clusters. For this reason, during implementation of this enhancement, no
+assumption that MicroShift will always run as a single-node should be made and potential multi-node
+should be kept in mind.
+
+### Drawbacks
+
+Multus does not actively watch NetworkAttachmentDefinitions or annotations therefore to make changes
+to these resources effective Pod must be re-created. This behavior is reasonable because it does not
+disrupt Pod's networking and should not be a problem in production environments where we expect to
+have stable configurations.
+
+Multus also does not observe the underlying bridge interfaces, therefore if one is rebuilt, Pod's
+interface might stop working (see [BZ #2066351](https://bugzilla.redhat.com/show_bug.cgi?id=2066351)).
+If these limitations are ever addressed (see [NP-606](https://issues.redhat.com/browse/NP-606) and 
+[NP-608](https://issues.redhat.com/browse/NP-608)), they would, most likely, be part of the thick
+Multus plugin.
+
+Even though these are limitation of the Multus itself, they should be documented.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Clean up of Multus artifacts on disk during system rollback or when `microshift-cleanup-data` is executed:
+   Should we invest in creating a lightweight plugin architecture to avoid adding cleanup of Multus
+   to MicroShift's core cleanup script? This architecture could be based on existing
+   [kubectl plugin](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) architecture
+   where plugins start with `kubectl-`, so we could use `microshift-cleanup-plugin-` prefix.
```

Comment:
Choice is between
- putting Multus' (non-failing in case of absence) cleanup in [microshift-cleanup-data script](https://github.com/openshift/microshift/blob/main/scripts/microshift-cleanup-data.sh) which creates a weak link in the wrong direction (main product being aware of optional addon)
- > providing some dynamic way for an add-on to extend the cleanup

Reply:
Your first option was what I meant by hard-coding, so I think we're on the same page.

I'm inclined to say we do want some plugin system. I'm not sure the kubectl pattern fits the best. We wouldn't want users to run a separate command, right? We want the main cleanup script to also run the cleanup steps for the plug-ins. Something like the pattern for init.d or greenboot fits better for that case.

Do we have to worry about ordering of plugins compared to the main microshift cleanup steps? Do we need a separate design document to describe how that would work?

Reply:
You're right, kubectl's plugin arch might not be best fit here.

> Do we have to worry about ordering of plugins compared to the main microshift cleanup steps? Do we need a separate design document to describe how that would work?

For multus cleanup, I think it doesn't matter too much, but it should be "bound" to `--ovn` option (which would also be executed on `--all`). And for now, I don't see other components that would need an additional cleanup (which essentially cleans up something on the host, not cluster).
That situation will change when we get to pluggable CNIs - I expect most of them to leave something on the host. So maybe for now we don't want to dive into this topic as we don't have concrete use case? Having couple of commands to remove Multus' stuff but not fail when they're not present seems good enough compared to the alternative which would definitely require design doc.

Reply:
If we can describe a process for the user to run 2 commands in the right order instead of just 1, that seems good enough for a first iteration. We can put "plugins for the cleanup scripts" on the backlog as a feature for a future release.